File: app/src/main/java/wxm/KeepAccount/ui/usr/FrgUsrAdd.java
package wxm.KeepAccount.ui.usr;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.Message;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.OnClick;
import wxm.KeepAccount.define.EMsgType;
import wxm.androidutil.FrgUtility.FrgSupportBaseAdv;
import wxm.androidutil.util.UtilFun;
import wxm.androidutil.util.WRMsgHandler;
import wxm.KeepAccount.R;
import wxm.KeepAccount.define.GlobalDef;
import wxm.KeepAccount.define.UsrItem;
import wxm.KeepAccount.utility.ContextUtil;

/**
 * add user
 * Created by WangXM on 2016/11/29.
 */
public class FrgUsrAdd extends FrgSupportBaseAdv
        implements TextView.OnEditorActionListener {
    // for ui
    @BindView(R.id.et_usr_name)
    EditText mETUsrName;
    @BindView(R.id.et_pwd)
    EditText mETPwd;
    @BindView(R.id.et_repeat_pwd)
    EditText mETRepeatPwd;
    // fro res
    @BindString(R.string.error_no_usrname)
    String mRSErrorNoUsrName;

    @BindString(R.string.error_invalid_password)
    String mRSErrorInvalidPWD;
    @BindString(R.string.error_repeatpwd_notmatch)
    String mRSErrorRepeatPwdNotMatch;
    // for data
    private LocalMsgHandler mMHHandler;

    @Override
    protected int getLayoutID() {
        return R.layout.vw_usr_add;
    }

    @Override
    protected boolean isUseEventBus() {
        return false;
    }

    @Override
    protected void initUI(Bundle bundle) {
        mMHHandler = new LocalMsgHandler(this);

        mETUsrName.setOnEditorActionListener(this);
        mETPwd.setOnEditorActionListener(this);
        mETRepeatPwd.setOnEditorActionListener(this);
    }


    @OnClick({R.id.bt_confirm, R.id.bt_giveup})
    void onSelfClick(View v) {
        switch (v.getId()) {
            case R.id.bt_confirm: {
                if (checkInput()) {
                    Intent data = new Intent();
                    data.putExtra(UsrItem.FIELD_NAME, mETUsrName.getText().toString());
                    data.putExtra(UsrItem.FIELD_PWD, mETPwd.getText().toString());

                    Message m = Message.obtain(ContextUtil.Companion.getMsgHandler(),
                            EMsgType.USR_ADD.getId());
                    m.obj = new Object[]{data, mMHHandler};
                    m.sendToTarget();
                }
            }
            break;

            case R.id.bt_giveup: {
                int ret_data = GlobalDef.INTRET_GIVEUP;

                Activity ac = getActivity();
                Intent data = new Intent();
                ac.setResult(ret_data, data);
                ac.finish();
            }
            break;
        }
    }

    @Override
    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
        switch (v.getId()) {
            case R.id.et_usr_name: {
                Log.d(LOG_TAG, "now usr name : " + mETUsrName.getText().toString());
            }
            break;

            case R.id.et_pwd: {
                Log.d(LOG_TAG, "now pwd : " + mETPwd.getText().toString());
            }
            break;

            case R.id.et_repeat_pwd: {
                Log.d(LOG_TAG, "now repeatPWD : " + mETRepeatPwd.getText().toString());
            }
            break;
        }

        return false;
    }


    /// PRIVATE BEGIN
    /**
     * clean UI
     */
    private void repeatInput() {
        mETUsrName.setText("");
        mETPwd.setText("");
        mETRepeatPwd.setText("");

        mETUsrName.requestFocus();
    }

    /**
     * check input data then give prompting message
     * @return  true if data legal else false
     */
    private boolean checkInput() {
        String usr_name = mETUsrName.getText().toString();
        String usr_pwd = mETPwd.getText().toString();
        String usr_r_pwd = mETRepeatPwd.getText().toString();

        boolean bret = true;
        if (UtilFun.StringIsNullOrEmpty(usr_name)) {
            mETUsrName.setError(mRSErrorNoUsrName);
            mETUsrName.requestFocus();
            bret = false;
        }

        if (bret && 4 > usr_pwd.length()) {
            mETPwd.setError(mRSErrorInvalidPWD);
            mETPwd.requestFocus();
            bret = false;
        }

        if (bret && !usr_pwd.equals(usr_r_pwd)) {
            mETRepeatPwd.setError(mRSErrorRepeatPwdNotMatch);
            mETRepeatPwd.requestFocus();
            bret = false;
        }

        return bret;
    }
    /// PRIVATE END

    private static class LocalMsgHandler extends WRMsgHandler<FrgUsrAdd> {
        LocalMsgHandler(FrgUsrAdd ac) {
            super(ac);
            TAG = "LocalMsgHandler";
        }

        @Override
        protected void processMsg(Message m, FrgUsrAdd home) {
            EMsgType et = EMsgType.Companion.getEMsgType(m.what);
            if(null == et)
                return;

            if(EMsgType.REPLAY == et)   {
                EMsgType et_inner = EMsgType.Companion.getEMsgType(m.arg1);
                if(null != et_inner)    {
                    if(EMsgType.USR_ADD == et_inner)    {
                        afterAddUsr(m, home);
                    }
                }
            } else  {
                Log.e(TAG, String.format("msg(%s) can not process", m.toString()));
            }
        }

        private void afterAddUsr(Message m, FrgUsrAdd home) {
            Activity ac = home.getActivity();
            Object[] arr = UtilFun.cast_t(m.obj);
            boolean ret = (boolean) arr[0];
            if (ret) {
                Intent data = UtilFun.cast_t(arr[1]);

                int ret_data = GlobalDef.INTRET_USR_ADD;
                ac.setResult(ret_data, data);
                ac.finish();
            } else {
                String sstr = "添加用户失败!";
                if (2 < arr.length)
                    sstr = UtilFun.cast(arr[2]);

                Toast.makeText(ContextUtil.Companion.getInstance(), sstr,
                        Toast.LENGTH_LONG).show();
                home.repeatInput();
            }
        }
    }
}

File: app/src/main/java/wxm/KeepAccount/ui/usr/FrgUsrAdd.kt
package wxm.KeepAccount.ui.usr

import android.content.Intent
import android.os.Bundle
import android.os.Message
import android.view.KeyEvent
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import kotterknife.bindView
import wxm.KeepAccount.R
import wxm.KeepAccount.define.EMsgType
import wxm.KeepAccount.define.GlobalDef
import wxm.KeepAccount.define.UsrItem
import wxm.KeepAccount.utility.ContextUtil
import wxm.androidutil.app.AppBase
import wxm.androidutil.log.TagLog
import wxm.androidutil.ui.frg.FrgSupportBaseAdv
import wxm.androidutil.ui.view.EventHelper
import wxm.androidutil.util.UtilFun
import wxm.androidutil.util.WRMsgHandler

/**
 * add user
 * Created by WangXM on 2016/11/29.
 */
class FrgUsrAdd : FrgSupportBaseAdv(), TextView.OnEditorActionListener {
    // for ui
    private val mETUsrName: EditText by bindView(R.id.et_usr_name)
    private val mETPwd: EditText by bindView(R.id.et_pwd)
    private val mETRepeatPwd: EditText by bindView(R.id.et_repeat_pwd)

    // for res
    private val mRSErrorNoUsrName: String = AppBase.getString(R.string.error_no_usr_name)
    private val mRSErrorInvalidPWD: String = AppBase.getString(R.string.error_invalid_password)
    private var mRSErrorRepeatPwdNotMatch: String = AppBase.getString(R.string.error_pwd_not_match)

    // for data
    private var mMHHandler: LocalMsgHandler = LocalMsgHandler(this)

    override fun getLayoutID(): Int {
        return R.layout.vw_usr_add
    }

    override fun isUseEventBus(): Boolean {
        return false
    }

    override fun initUI(bundle: Bundle?) {
        if (null == bundle) {
            mETUsrName.setOnEditorActionListener(this)
            mETPwd.setOnEditorActionListener(this)
            mETRepeatPwd.setOnEditorActionListener(this)

            EventHelper.setOnClickListener(view!!,
                    intArrayOf(R.id.bt_confirm, R.id.bt_giveup),
                    View.OnClickListener { v ->
                        when (v.id) {
                            R.id.bt_confirm -> {
                                if (checkInput()) {
                                    val data = Intent().apply {
                                        putExtra(UsrItem.FIELD_NAME, mETUsrName.text.toString())
                                        putExtra(UsrItem.FIELD_PWD, mETPwd.text.toString())
                                    }

                                    Message.obtain(ContextUtil.msgHandler,
                                            EMsgType.USR_ADD.id).let {
                                        it.obj = arrayOf(data, mMHHandler)
                                        it.sendToTarget()
                                    }
                                }
                            }

                            R.id.bt_giveup -> {
                                activity.let {
                                    it.setResult(GlobalDef.INTRET_GIVEUP, Intent())
                                    it.finish()
                                }
                            }
                        }
                    })
        }
    }

    override fun onEditorAction(v: TextView, actionId: Int, event: KeyEvent): Boolean {
        when (v.id) {
            R.id.et_usr_name -> {
                TagLog.d("now usr name : " + mETUsrName.text.toString())
            }

            R.id.et_pwd -> {
                TagLog.d("now pwd : " + mETPwd.text.toString())
            }

            R.id.et_repeat_pwd -> {
                TagLog.d("now repeatPWD : " + mETRepeatPwd.text.toString())
            }
        }

        return false
    }


    /// PRIVATE BEGIN
    /**
     * clean UI
     */
    private fun repeatInput() {
        mETUsrName.setText("")
        mETPwd.setText("")
        mETRepeatPwd.setText("")

        mETUsrName.requestFocus()
    }

    /**
     * check input data then give prompting message
     * @return  true if data legal else false
     */
    private fun checkInput(): Boolean {
        val usrName = mETUsrName.text.toString()
        val usrPwd = mETPwd.text.toString()
        val usrRPwd = mETRepeatPwd.text.toString()

        if (UtilFun.StringIsNullOrEmpty(usrName)) {
            mETUsrName.apply {
                error = mRSErrorNoUsrName
                requestFocus()
            }

            return false
        }

        if (4 > usrPwd.length) {
            mETPwd.apply {
                error = mRSErrorInvalidPWD
                requestFocus()
            }

            return false
        }

        if (usrPwd != usrRPwd) {
            mETRepeatPwd.apply {
                error = mRSErrorRepeatPwdNotMatch
                requestFocus()
            }

            return false
        }

        return true
    }
    /// PRIVATE END

    private class LocalMsgHandler internal constructor(ac: FrgUsrAdd) : WRMsgHandler<FrgUsrAdd>(ac) {
        override fun processMsg(m: Message, home: FrgUsrAdd) {
            when(EMsgType.getEMsgType(m.what))  {
                EMsgType.REPLAY -> {
                    EMsgType.getEMsgType(m.arg1)?.let {
                        if (EMsgType.USR_ADD === it) {
                            afterAddUsr(m, home)
                        }
                    }
                }

                else  -> {
                    TagLog.e("msg($m) can not process")
                }
            }
        }

        private fun afterAddUsr(m: Message, home: FrgUsrAdd) {
            val arr = UtilFun.cast_t<Array<Any>>(m.obj)
            if (arr[0] as Boolean) {
                home.activity.apply {
                    setResult(GlobalDef.INTRET_USR_ADD, arr[1] as Intent)
                    finish()
                }
            } else {
                val sstr = if (2 < arr.size) UtilFun.cast(arr[2])
                else "添加用户失败!"

                Toast.makeText(ContextUtil.self, sstr, Toast.LENGTH_LONG).show()
                home.repeatInput()
            }
        }
    }
}
